File: site/pages/stats/DefenseAlgorithm.js
import Algorithm from './Algorithm'
import {POKEMON_LIST} from "app/data/Database";

class DefenseAlgorithm extends Algorithm {
  name = "defense";

  compute(pokemon, quickMove, chargeMove, typeEffectiveness = { }) {
    const quickEffectiveness = typeEffectiveness[quickMove.type] || 1;
    const chargeEffectiveness = typeEffectiveness[chargeMove.type] || 1;
    const moveRating = quickMove.pow * 8 * quickEffectiveness
      + chargeMove.pow * (chargeMove.bars * 0.5 + 0.5) * chargeEffectiveness;
    return moveRating * pokemon.getAttack() * pokemon.getDefense() * pokemon.getStamina()
  }
}

const inst = new DefenseAlgorithm();
inst.rank(POKEMON_LIST);
export default inst

File: src/com/sucy/go/ranks/impl/DefenseAlgorithm.java
package com.sucy.go.ranks.impl;

import com.sucy.go.Move;
import com.sucy.go.Pokemon;
import com.sucy.go.ranks.RankAlgorithm;

/**
 * PokemonGo © 2018
 * com.sucy.go.ranks.impl.DefenseAlgorithm
 */
public class DefenseAlgorithm extends RankAlgorithm {
    @Override
    public String getName() {
        return "defenders";
    }

    @Override
    public int compute(
            final Pokemon pokemon, final Move quickMove, final Move chargeMove) {

        double moveRating = 0.1 * (quickMove.pow * 5 + chargeMove.pow * chargeMove.bars);
        return (int)(moveRating * pokemon.attack() * pokemon.defense() * pokemon.stamina());
    }
}
